### PR TITLE
add and support "ReplaceGraveyardValid" for scholar_of_the_lost_trove

### LINKED ISSUE #2105: Scholar of the Lost Trove: missing exception for artifacts
> If you cast an artifact spell this way, it won’t be exiled if it would be put into your graveyard this turn. 

While fixing that it might also be good to see if the effect can be removed earlier now, it looks like all cards had the "this turn" part removed now anyway since it didn't make sense...

### PR DESCRIPTION
**Pull request by Northmoc**
closes #2105

### CODE REVIEW

**Comment by Hanmac** [comment]
Does this handle cast from different states yet? Like MDFC where the back is an Instant or sorcery?

**Comment by Northmoc** [comment]
> Does this handle cast from different states yet? Like MDFC where the back is an Instant or sorcery?

Looks like no – seems like a deeper issue with the PlayEffect?

**Comment by tool4ever** [comment]
No, it should work
but in your example you must target first and Blex just isn't a legal one ;)

**Comment by Hanmac** [comment]
Some Adventure cards are Artifact:
* Altar of Bhaal
* Ghost Lantern
* Monster Manual
* Horn of Valhalla
* Two-Handed Axe

Pestilent Cauldron is a MDFC (even if Restorative Burst already exiles)

File: forge-game/src/main/java/forge/game/ability/effects/PlayEffect.java [hunk 1]
```diff
@@ -407,7 +407,10 @@ public void resolve(final SpellAbility sa) {
 
             // can't be done later
             if (sa.hasParam("ReplaceGraveyard")) {
-                addReplaceGraveyardEffect(tgtCard, sa, sa.getParam("ReplaceGraveyard"), moveParams);
+                if (!sa.hasParam("ReplaceGraveyardValid")
+                        || tgtCard.isValid(sa.getParam("ReplaceGraveyardValid").split(","), activator, source, sa)) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/ability/effects/PlayEffect.java:411]: Just run the isValid against tgtSA, that will have the right state

**Review by tool4ever — APPROVED** [review]

**Comment by Hanmac** [comment]
one thing that's still bugging me is that it's optional ... meaning that if you don't choose to cast the card after targeting it, does the ReplaceGraveyard Effect still continue to exist? 🤔

**Comment by Northmoc** [comment]
> one thing that's still bugging me is that it's optional ... meaning that if you don't choose to cast the card after targeting it, does the ReplaceGraveyard Effect still continue to exist? 🤔

I tried this with a few different scenarios... including using Wrenn and Six emblem to give the targeted but not cast card retrace and it seems the ReplaceGraveyard Effect does not apply. So I think we're safe for now?

**Pull request merged**